### PR TITLE
Aggressively prune specialize() branches

### PR DESCRIPTION
**Pull request by steven-johnson**
When a specialize() condition is const-false, we should always prune it
away before lowering.

When a specialize() condition is const-true, we should always prune
away all subsequent specializations (except for those that are also
const-true).

(Note that this pruning already effectively happened later during
simplification, but pruning earlier removes a nontrivial amount of
work.)

### CODE REVIEW

File: src/SimplifySpecializations.cpp [hunk 1]
```diff
@@ -111,8 +141,110 @@ vector<Definition> propagate_specialization_in_definition(Definition &def) {
...
+        } else {
+            vector_store_lanes = 0;
+        }
     }
+    return 0;
+}
+
+}  // namespace
+
+void simplify_specializations_test() {
```
Comment (abadams) [inline src/SimplifySpecializations.cpp]: Even though this uses Internal apis, I think it would be better as a standalone JIT test

**Comment by steven-johnson** [comment]
Changes made per comments.

File: test/correctness/specialize.cpp [hunk 2]
```diff
@@ -519,6 +522,92 @@ int main(int argc, char **argv) {
...
+    }
+    {
+        Var x, y;
+        Param<int> p;
+        Expr const_false = Expr(0) == Expr(1);
+        Expr const_true = Expr(0) == Expr(0);
+        Expr different_const_true = Expr(1) == Expr(1);
+
+        // Check that we aggressively prune all specializations after specialize(const-true)
+        // except for other occurrences of specialize(const-true)
```
Comment (abadams) [inline test/correctness/specialize.cpp]: Out-of-date comment

**Comment by steven-johnson** [comment]
Added a subsequent check to convert a trailing specialize(true) -> replace the default schedule (and append sub-specializations). I *think* I did this correctly, but @psuriana should probably review commit e00f15f before this lands.

File: test/correctness/specialize.cpp [hunk 3]
```diff
@@ -589,10 +590,11 @@ int main(int argc, char **argv) {
         simplify_specializations(env);
 
         const auto &s = f.function().definition().specializations();
-        _halide_user_assert(s.size() == 2);
+        // Note that this is 1 (rather than 2) because the final const-true
+        // Specialization will be hoisted into the main Schdule.
```
Comment (abadams) [inline test/correctness/specialize.cpp]: Schdule -> Schedule

File: test/correctness/specialize.cpp [hunk 4]
```diff
@@ -519,6 +522,93 @@ int main(int argc, char **argv) {
...
+        vector_store_lanes = 0;
+        p.set(42);  // just a nonzero value
+        f.realize(100);
+        _halide_user_assert(vector_store_lanes == 4);
+    }
+
+    {
+        Var x, y;
+        Param<int> p;
+        Expr const_false = Expr(0) == Expr(1);
```
Comment (psuriana) [inline test/correctness/specialize.cpp]: You could probably use const_true() and const_false() instead

Comment (psuriana) [inline test/correctness/specialize.cpp]: Ah I see, so you use it to trigger different specialization. nvm.

**Comment by psuriana** [comment]
LGTM

**Comment by abadams** [comment]
A legitimate failure was caught by the new buildbot setup! Hooray!

Looks like there's a missing EXPORT qualifier on simplify_specializations.

**Comment by steven-johnson** [comment]
> missing EXPORT

It was already missing, so this shouldn't be a new failure. (I'll add it anyway.)

**Pull request merged**